File: Dynamic Programming/Book_Shop.py
import sys,math,random
from heapq import heappush,heappop
from bisect import bisect_right,bisect_left
from collections import Counter,deque,defaultdict
from itertools import permutations

# functions #
MOD = 998244353
MOD = 10**9 + 7
RANDOM = random.randrange(2**62)
def gcd(a,b):
    if a%b==0:
        return b
    else:
        return gcd(b,a%b)
def lcm(a,b):
    return a//gcd(a,b)*b
def w(x):
    return x ^ RANDOM
##

#String hashing : sh, fenwick sortedlist : fsortl, Number : numtheory, SparseTable : SparseTable
#bucket sorted list : bsortl, segment tree(lazy propogation) : SegmentTree, bootstrap : bootstrap
#binary indexed tree : BIT, segment tree(point updates) : SegmentPoint, Convex Hull : hull
#Combinatorics : pnc, Diophantine Equations : dpheq
#Template : https://github.com/OmAmar106/Template-for-Competetive-Programming

def solve():
    n,X = list(map(int, sys.stdin.readline().split()))
    L1 = list(map(int, sys.stdin.readline().split()))
    L2 = list(map(int, sys.stdin.readline().split()))

    dp = [0 for i in range(X+1)]
 
    for i in range(n):
        for j in range(X,L1[i]-1,-1):
            dp[j] = max(dp[j],dp[j-L1[i]]+L2[i])

    print(dp[X])

    #st = sys.stdin.readline().strip()
solve()

File: Dynamic Programming/Book_Shop.cpp
// Choose cpp when the time
// limit will for sure give tle in pypy
#include <bits/stdc++.h>
using namespace std;
#define int long long
#define INF LLONG_MAX
#define fastio ios::sync_with_stdio(false); cin.tie(0);
#define print(arr) for (auto it : arr){cout<<it<<" ";}cout<<endl;
#define len(arr) arr.size()
#define printf(x) cout << x << endl;
#define printm(map) cout<<"{";for(auto it: map){cout<<it.first<<":"<<it.second<<",";};cout<<"}"<<endl;

// const int MOD = 998244353;
const int MOD = 1e9 + 7;

int gcd(int a, int b) {
    if (b == 0) return a;
    return gcd(b, a % b);
}

int lcm(int a, int b) {
    return a / gcd(a, b) * b;
}

// String hashing: sh/shclass, Number: numtheory, SparseTable: SparseTable
// Segment Tree(lazy propogation): SegmentTree
// binary indexed tree: BIT, Segment Tree(point updates): SegmentPoint, Convex Hull: hull, Trie/Treap: Tries
// Combinatorics: pnc, Diophantine Equations: dpheq, Graphs: graphs, DSU: DSU, Geometry: Geometry, FFT: fft
// Persistent Segment Tree: perseg, FreqGraphs: bgraph
// Template : https://github.com/OmAmar106/Template-for-Competetive-Programming

void solve() {
    int n,x; cin >> n>>x;
    vector<int> L(n);
    for (int &x : L) cin >> x;
    vector<int> L1(n);
    for (int &x : L1) cin >> x;
    int dp[x+1] = {-INF};
    dp[0] = 0;
    int ans = 0;
    for(int i=0;i<n;i++){
        for(int j=x-L[i];j>=0;j--){
            dp[j+L[i]] = max(dp[j+L[i]],dp[j]+L1[i]); 
        }
    }
    for(int i=0;i<x+1;i++){
        ans = max(ans,dp[i]);
    }
    printf(ans);
}

int32_t main() {
    fastio
    int t=1;
    while (t--) solve();
    return 0;
}